File: analyze_commit.js
const fs = require('fs');
const zlib = require('zlib');
const path = require('path');

const baseDir = r`D:\Dropbox\DX\Twitch\CodeRushed\MrAnnouncerBot`;
process.chdir(baseDir);

const COMMIT = 'be0e1ceec39a7caf0b26b579a22762f2109a7cd8';

function readGitObject(sha) {
    const objPath = path.join('.git', 'objects', sha.substring(0, 2), sha.substring(2));
    if (!fs.existsSync(objPath)) {
        return [null, null];
    }
    const compressed = fs.readFileSync(objPath);
    const decompressed = zlib.uncompressSync(compressed);
    const nullIdx = decompressed.indexOf(0);
    const header = decompressed.toString('ascii', 0, nullIdx);
    const objType = header.split(' ')[0];
    const content = decompressed.subarray(nullIdx + 1);
    return [objType, content];
}

function parseCommit(content) {
    const text = content.toString('utf-8', { errors: 'replace' });
    const lines = text.split('\n');
    const result = {};
    let msgLines = [];
    let inMsg = false;
    
    for (const line of lines) {
        if (inMsg) {
            msgLines.push(line);
        } else if (line === '') {
            inMsg = true;
        } else if (line.startsWith('tree ')) {
            result.tree = line.substring(5).trim();
        } else if (line.startsWith('parent ')) {
            result.parent = line.substring(7).trim();
        } else if (line.startsWith('author ')) {
            result.author = line.substring(7).trim();
        } else if (line.startsWith('committer ')) {
            result.committer = line.substring(10).trim();
        }
    }
    result.message = msgLines.join('\n').trim();
    return result;
}

console.log('='.repeat(80));
console.log(`GIT COMMIT ANALYSIS: ${COMMIT}`);
console.log('='.repeat(80));

const [objType, content] = readGitObject(COMMIT);
if (content === null) {
    console.log(`ERROR: Could not find commit object ${COMMIT}`);
    process.exit(1);
}

console.log(`\nObject type: ${objType}`);
const commitData = parseCommit(content);

console.log(`Tree:       ${commitData.tree || 'N/A'}`);
console.log(`Parent:     ${commitData.parent || 'N/A'}`);
console.log(`Author:     ${commitData.author || 'N/A'}`);
console.log(`Committer:  ${commitData.committer || 'N/A'}`);
console.log(`Message:    ${commitData.message || 'N/A'}`);

File: analyze_commit.py
#!/usr/bin/env python3
"""Read git commit object and show what changed"""
import zlib
import os
import struct
import hashlib
import subprocess

os.chdir(r'D:\Dropbox\DX\Twitch\CodeRushed\MrAnnouncerBot')

commit_hash = 'be0e1ceec39a7caf0b26b579a22762f2109a7cd8'
obj_path = '.git/objects/be/0e1ceec39a7caf0b26b579a22762f2109a7cd8'

print("=" * 70)
print("READING GIT COMMIT OBJECT")
print("=" * 70)
print(f"\nCommit hash: {commit_hash}")
print(f"Object path: {obj_path}")
print(f"File exists: {os.path.exists(obj_path)}")

if os.path.exists(obj_path):
    print("\n" + "=" * 70)
    print("DECOMPRESSED COMMIT OBJECT CONTENT")
    print("=" * 70)
    
    with open(obj_path, 'rb') as f:
        compressed = f.read()
        decompressed = zlib.decompress(compressed)
        content = decompressed.decode('utf-8', errors='replace')
        print(content)
        
        # Parse the commit object to extract parent and tree
        print("\n" + "=" * 70)
        print("PARSED COMMIT DETAILS")
        print("=" * 70)
        
        lines = content.split('\n')
        tree_hash = None
        parent_hash = None
        author = None
        committer = None
        message = None
        
        # Parse header
        for i, line in enumerate(lines):
            if line.startswith('tree '):
                tree_hash = line.replace('tree ', '').strip()
                print(f"\nTree hash: {tree_hash}")
            elif line.startswith('parent '):
                parent_hash = line.replace('parent ', '').strip()
                print(f"Parent hash: {parent_hash}")
            elif line.startswith('author '):
                author = line
                print(f"Author: {author}")
            elif line.startswith('committer '):
                committer = line
                print(f"Committer: {committer}")
            elif line == '' and i > 0:
                message = '\n'.join(lines[i+1:])
                print(f"\nCommit message:\n{message}")
                break

print("\n" + "=" * 70)
print("USING GIT COMMANDS FOR DETAILED CHANGE INFORMATION")
print("=" * 70)

# Try to use git commands to get the actual diff
try:
    print("\n--- GIT SHOW --STAT ---")
    result = subprocess.run(
        ['git', 'show', commit_hash, '--stat'],
        capture_output=True,
        text=True,
        timeout=10
    )
    print(result.stdout)
    if result.stderr:
        print("STDERR:", result.stderr)
except Exception as e:
    print(f"Error running git show --stat: {e}")

try:
    print("\n--- GIT SHOW --NAME-ONLY ---")
    result = subprocess.run(
        ['git', 'show', commit_hash, '--name-only'],
        capture_output=True,
        text=True,
        timeout=10
    )
    print(result.stdout)
    if result.stderr:
        print("STDERR:", result.stderr)
except Exception as e:
    print(f"Error running git show --name-only: {e}")

try:
    print("\n--- GIT DIFF-TREE ---")
    result = subprocess.run(
        ['git', 'diff-tree', '--no-commit-id', '-r', commit_hash],
        capture_output=True,
        text=True,
        timeout=10
    )
    print(result.stdout)
    if result.stderr:
        print("STDERR:", result.stderr)
except Exception as e:
    print(f"Error running git diff-tree: {e}")
